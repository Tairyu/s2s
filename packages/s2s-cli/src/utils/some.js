// @flow
import micromatch from 'micromatch'

export default function(list: string | string[], patterns: string | string[]) {
  list = [].concat(list)
  for (let i = 0; i < list.length; i++) {
    if (micromatch(list[i], patterns).length === 1) {
      return true
    }
  }
  return false
}
