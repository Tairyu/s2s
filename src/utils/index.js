// @flow
import path from 'path'
import fs from 'fs'
import { transformFileSync } from 'babel-core'

type Path = string
type Filename = string
type Code = string

export function log(msg: string) {
  console.log(msg)
}

export function compile(
  filename: Filename,
  opts: Object
): { code?: Code, ignored: boolean } {
  try {
    return transformFileSync(filename, opts)
  } catch (err) {
    console.error(toErrorStack(err))
    return { ignored: true }
  }
}

export function toErrorStack(err: Object) {
  if (err._babel && err instanceof SyntaxError) {
    return `${err.name}: ${err.message}\n${err.codeFrame}`
  }
  return err.stack
}

export function getOutputPath(output: Path, input: Path): Path {
  if (path.isAbsolute(output)) {
    return output
  }

  return path.resolve(path.dirname(input), output)
}

export function writeFileSync(outputPath: Path, code: string) {
  fs.writeFileSync(outputPath, code, 'utf-8')
}

export function relativeFromCwd(input: Path) {
  return path.relative(process.cwd(), input)
}

export function getDirAndBaseName(testPath: string) {
  const dirname = path.dirname(testPath)
  const basename = path.basename(testPath)

  return {
    dirname,
    basename,
  }
}
