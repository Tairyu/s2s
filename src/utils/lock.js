// @flow

export default class Lock {
  _lock: Set<string>
  _timer: ?number

  constructor() {
    this._lock = new Set()
  }

  has(input: string) {
    return this._lock.has(input)
  }

  add(input: string) {
    this.clear()
    clearTimeout(this._timer)

    this._lock.add(input)

    this._timer = setTimeout(() => {
      this._lock.delete(input)
    }, 1000)
  }

  clear() {
    this._lock.clear()
  }
}
