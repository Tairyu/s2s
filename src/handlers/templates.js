// @flow
import path from 'path'
import chalk from 'chalk'
import cpFile from 'cp-file'
import type { Path, Template } from '../types'
import { formatText, trimAndFormatPath } from '../reporters'
import { log, isAlreadyExist } from '../utils'

function handleCopyError(err: Error & { path: string, code: string }) {
  if (err.name === 'CpFileError' && err.code === 'ENOENT') {
    const errorText = `${chalk.reset.inverse.bold.red(
      'TEMPLATE'
    )} ${trimAndFormatPath(err.path)}
          no such file or directory`
    log(errorText)
  } else {
    log(err.stack)
  }
}

const DEFAULT_TEMPLATES_DIR = 'templates'

function handleTemplate(input: Path, template: Template, templatesDir: string) {
  if (!template.test.test(input)) {
    return
  }

  if (isAlreadyExist(input)) {
    return
  }

  const templatePath = path.join(templatesDir, template.input)
  cpFile.sync(templatePath, input)

  log(formatText('TEMPLATE', templatePath, input))
}

export default function handleTemplates(
  input: Path,
  templates: Template[] = [],
  templatesDir: string = DEFAULT_TEMPLATES_DIR
) {
  for (const template of templates) {
    try {
      handleTemplate(input, template, templatesDir)
    } catch (err) {
      handleCopyError(err)
    }
  }
}
