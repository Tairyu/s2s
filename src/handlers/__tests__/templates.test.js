// @flow
import { resolve, join } from 'path'
import rimraf from 'rimraf'
import cpFile from 'cp-file'
import * as utils from '../../utils'
import * as templates from '../templates'

let writeSpy
let logSpy

const fn = x => x

beforeEach(() => {
  rimraf.sync(resolve(__dirname, 'fixtures', 'copy-result.js'))
  logSpy = jest.spyOn(utils, 'log').mockImplementation(fn)
  writeSpy = jest.spyOn(utils, 'writeFileSync').mockImplementation(fn)
})

afterEach(() => {
  logSpy.mockRestore()
  writeSpy.mockRestore()
})

const fixturesPath = resolve(__dirname, 'fixtures')
const getEventPath = (...x) => join(fixturesPath, ...x)
const eventPath = getEventPath('copy-result.js')

test('called log with output result', () => {
  const template = { test: /copy-result.js/, input: 'copy.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls[0][0]).toMatchSnapshot()
})

test('already exist', () => {
  const template = { test: /copy-result.js/, input: 'copy.js' }

  templates.default(eventPath, [template], fixturesPath)
  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls.length).toBe(1)
})

test('handle copy error', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls[0][0]).toMatch('no such file or directory')
})

test('handle nomal error', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  const cpFIleSpy = jest.spyOn(cpFile, 'sync').mockImplementation(() => {
    throw new Error('hello')
  })
  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls[0][0]).toMatch('hello')
  cpFIleSpy.mockRestore()
})

test('template not found', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).toHaveBeenCalled()
  expect(logSpy.mock.calls[0][0]).toMatch('no such file or directory')
})

test('test regex not match', () => {
  const template = { test: /not-found.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).not.toHaveBeenCalled()
})

test('no result when templates = []', () => {
  templates.default(eventPath, [])
  expect(logSpy).not.toHaveBeenCalled()
})

test('no result when templates = undefined', () => {
  templates.default(eventPath)
  expect(logSpy).not.toHaveBeenCalled()
})
