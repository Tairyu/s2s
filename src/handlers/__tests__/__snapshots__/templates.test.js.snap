// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`called log with output result 1`] = `"TEMPLATE src/handlers/__tests__/fixtures/copy.js â†’ src/handlers/__tests__/fixtures/copy-result.js"`;
